Upgrade to v6 major

danielrbradley opened the pull request:
Fixes https://github.com/pulumi/pulumi-azuread/issues/1416

Adds an upgrade test for the application resource.

## Migration guide

The property `endDateRelative` has deprecated in favor of `endDate`. If using `endDateRelative` with a custom `startDate`, calculate the `endDate` manually and this will avoid requiring replacement of the resource:
- `azuread.ApplicationCertificate`
- `azuread.ApplicationPassword`
- `azuread.ServicePrincipalCertificate`
- `azuread.ServicePrincipalPassword`

The following deprecated properties are being removed:
- [azuread.Application](https://www.pulumi.com/registry/packages/azuread/api-docs/application): `applicationId` removed in favor of `client_id`. This won't cause replacement.
- [azuread.ApplicationCertificate](https://www.pulumi.com/registry/packages/azuread/api-docs/applicationcertificate/): `application_object_id` replaced with `application_id`. This will cause replacement.
- [azuread.ApplicationFederatedIdentityCredential](https://www.pulumi.com/registry/packages/azuread/api-docs/applicationfederatedidentitycredential/): `application_object_id` replaced with `application_id`. This will cause replacement.
- [azuread.ApplicationPassword](https://www.pulumi.com/registry/packages/azuread/api-docs/applicationpassword/): `application_object_id` replaced with `application_id`. This will cause replacement.
- [azuread.ApplicationPreAuthorized](https://www.pulumi.com/registry/packages/azuread/api-docs/applicationpreauthorized/): `application_object_id` replaced with `application_id`. This will cause replacement.
- [azuread.DirectoryRoleAssignment](https://www.pulumi.com/registry/packages/azuread/api-docs/directoryroleassignment/) `application_object_id` replaced with `application_id`. This will cause replacement.

thomas11 commented on sdk/go/azuread/applicationPassword.go:
```diff
@@ -151,15 +153,13 @@ func GetApplicationPassword(ctx *pulumi.Context,
 	EndDate *string `pulumi:"endDate"`
 	// A relative duration for which the password is valid until, for example `240h` (10 days) or `2400h30m`. Changing this field forces a new resource to be created.
+	//
+	// Deprecated: The `endDateRelative` property is deprecated and will be removed in a future version of the AzureAD provider. Please instead use the Terraform `timeadd()` function to calculate a value for the `endDate` property.
```
Unfortunate comment for Pulumi users since there is no `timeadd`.

danielrbradley replied:
So, after some digging, this value is actually coming from the deprecation notice in the code and not from the docs, so it's much harder to replace.

danielrbradley replied:
On attempting to override this message via the ProviderInfo, it turns out that if the upstream provider has a deprecation message, that will be preferred over the locally provided string, so this is impossible to set.

danielrbradley replied:
Bridged reference:
https://github.com/pulumi/pulumi-terraform-bridge/blob/f0fee4fb1db95e3aa8ebc505ab50c8152940c2d1/pkg/tfgen/generate.go#L636-L646

thomas11 commented:
Some questions:

> The property endDateRelative has deprecated in favor of endDate. This will cause replacement
If it's only deprecated and not removed, why does it cause replacement?

The `application_object_id` -> `application_id` rename should be a pretty straightforward state transform to spare our users the replacements. I understand the resources are all just metadata and should be quick to recreate, but GUIDs and passwords might change and cause disruption.

danielrbradley commented:
> The `application_object_id` -> `application_id` rename should be a pretty straightforward state transform to spare our users the replacements. I understand the resources are all just metadata and should be quick to recreate, but GUIDs and passwords might change and cause disruption.

I've now tested this for ApplicationPassword and the upstream provider already appears to have a migration in place as long as the calculated `endDate` matches what the `endDateReletive` resolved to in relation to the `startDate`.

thomas11 commented:
> > The `application_object_id` -> `application_id` rename should be a pretty straightforward state transform to spare our users the replacements. I understand the resources are all just metadata and should be quick to recreate, but GUIDs and passwords might change and cause disruption.
> 
> I've now tested this for ApplicationPassword and the upstream provider already appears to have a migration in place as long as the calculated `endDate` matches what the `endDateReletive` resolved to in relation to the `startDate`.

Good, but my comment you've quoted was about `application_id`, not `endDate`?

danielrbradley commented:
> Good, but my comment you've quoted was about `application_id`, not `endDate`?

🤦 yes

Changed the test to test the `applicationObjectId` -> `applicationId` which is the critical one because it's actually being removed. Also, fixed the update of the source which was missing. 

Also, having fixed the source edit step, the `endDateRelative` migration is actually not implemented yet, but the `applicationObjectId` is. I don't think this is a problem as the `endDateRelative` is still there for the time being, likely until the next major version, and it's likely a migration will be added upstream by then.

thomas11 reviewed (approved):
LGTM

danielrbradley commented:
Upgraded bridge and codegen and rebased before release.
